File: src/main/java/Executor.java
import burp.api.montoya.MontoyaApi;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonSyntaxException;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;

public final class Executor {

  public static ExecutorOutput execute(
      MontoyaApi api, String action, String source, HashMap<String, String> request) {
    StringBuilder output = new StringBuilder();
    String decodedOutput = "";
    String scriptToExecute;
    String command;
    StringBuilder stdErr = new StringBuilder();
    StringBuilder error = new StringBuilder();

    int numberOfOutputLines = 0;
    ExecutorOutput response = new ExecutorOutput();

    request.put("action", action);

    if ("request".equals(source)) {
      scriptToExecute = api.persistence().extensionData().getString(
          Constants.REQUEST_SCRIPT_PATH_KEY);
    } else {
      scriptToExecute = api.persistence().extensionData().getString(
          Constants.RESPONSE_SCRIPT_PATH_KEY);
    }

    command = Utils.getCommandFromPath(api.persistence(), scriptToExecute);

    if (!Utils.checkFileExists(scriptToExecute)) {
      response.setError(scriptToExecute + " is not a file");
      return response;
    }

    if (command == null) {
      response.setError("The selected script: "
          + scriptToExecute
          + " does not have a valid extension");
      return response;
    }

    File temp = null;

    try {
      temp = File.createTempFile("stripper_", ".json");

      try (Writer writer = new FileWriter(temp, StandardCharsets.UTF_8)) {
        Gson gson = new GsonBuilder().create();
        gson.toJson(request, writer);
      }
      File workingPath = new File(scriptToExecute);
      ProcessBuilder processBuilder =
          new ProcessBuilder(command, scriptToExecute, temp.getAbsolutePath());

      processBuilder.redirectErrorStream(false);
      processBuilder.directory(workingPath.getParentFile());

      Process process = processBuilder.start();

      BufferedReader reader = new BufferedReader(
          new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8)
      );

      BufferedReader errorReader = new BufferedReader(
          new InputStreamReader(process.getErrorStream(), StandardCharsets.UTF_8)
      );

      String line;


      while ((line = reader.readLine()) != null) {
        output.append(line);
        numberOfOutputLines += 1;
      }

      while ((line = errorReader.readLine()) != null) {
        stdErr.append(line + "\n");
      }

      decodedOutput =
          new String(
              api.utilities().base64Utils().decode(output.toString()).getBytes(),
              StandardCharsets.UTF_8
          );

      if (!temp.delete()) {
        api.logging().logToError("Could not delete: " + temp.getAbsolutePath());
      }

      if (decodedOutput.isEmpty()) {
        response.setError("Script's output was empty or null");
        response.setStdErr(stdErr.toString());
        return  response;
      }

      response = new Gson().fromJson(decodedOutput, ExecutorOutput.class);
      response.setStdErr(stdErr.toString());

      if (response.getEventLog() != null && !response.getEventLog().isEmpty()) {
        api.logging().raiseInfoEvent(response.getEventLog());
      }

      return response;
    } catch (IOException  | IllegalStateException | JsonSyntaxException |
        IllegalArgumentException e) {
      response.setError(String.format(
          Constants.STRIPPER_ERROR_TEMPLATE,
          command, scriptToExecute, "", e.toString(), decodedOutput));

      error.append("--------- Extension errors ---------" + "\n");
      error.append("[+] " + e.toString() + "\n");

      if (numberOfOutputLines > 1) {
        error.append("\n[+] more than one line detected in stdout,"
            + "please make sure you are not using console.log/print"
            + "debug only by printing to stderr");
      }

      if (temp != null) {
        if (!temp.delete()) {
          api.logging().logToError("Could not delete: " + temp.getAbsolutePath());
        }
      }

      response.setStdErr(stdErr.toString());
      response.setError(error.toString());
      return  response;
    }
  }
}

File: src/main/java/Executor.kt
import burp.api.montoya.MontoyaApi
import kotlinx.serialization.json.Json
import models.ExecutorOutput
import models.JsonRequestResponse
import java.io.File
import java.nio.charset.StandardCharsets
import java.util.*
import kotlin.io.path.createTempFile

object Executor {
  fun execute(
    montoyaApi: MontoyaApi,
    source: String,
    requestResponse: JsonRequestResponse
  ): ExecutorOutput {

    val script = if (source == "request") {
      montoyaApi.persistence().extensionData().getString(K.KEYS.REQUEST_SCRIPT_PATH) ?: ""
    } else {
      montoyaApi.persistence().extensionData().getString(K.KEYS.RESPONSE_SCRIPT_PATH) ?: ""
    }

    if (!utils.checkFileExists(script)) {
      return ExecutorOutput(error = "$script not a file")
    }

    val command = utils.getCommandFromPath(montoyaApi.persistence(), script)

    if (command == null) {
      return ExecutorOutput(
        error = "The selected script: "
          + script
          + " does not have a valid extension"
          + " or no binary selected"
      )
    }

    var temp: File? = null

    temp = createTempFile(prefix = "stripper_", suffix = ".json").toFile()
    val format = Json { encodeDefaults = true }
    val json = format.encodeToString(JsonRequestResponse.serializer(), requestResponse)
    temp.writeText(json, StandardCharsets.UTF_8)

    montoyaApi.logging().logToOutput("$command $script ${temp.absolutePath}")

    val process = ProcessBuilder(command, script, temp.absolutePath)
      .redirectErrorStream(false)
      .directory(File(script).parentFile)
      .start()

    val poutput = process.inputStream
      .bufferedReader(StandardCharsets.UTF_8)
      .use { it.readText() }

    val perror = process.errorStream
      .bufferedReader(StandardCharsets.UTF_8)
      .use { it.readText() }

    val decoded = try {
      String(
        Base64.getDecoder().decode(poutput.trim()),
        StandardCharsets.UTF_8
      )
    } catch (e: Exception) {
      return ExecutorOutput(error = "Error decoding base64. Don't use console.log()")
    }

    if (decoded.isBlank()) {
      return ExecutorOutput(
        error = "Script's output empty or null",
        stdErr = perror)
    }

    try {
      val o = Json.decodeFromString<ExecutorOutput>(decoded)

      if (!o.eventLog.isNullOrEmpty())
        montoyaApi.logging().raiseInfoEvent(o.eventLog)
      return o.copy(stdErr = perror)
    } catch (e: Exception) {
      montoyaApi.logging().logToOutput(e.toString())

      val error = StringBuilder()
        .append("--------- Extension errors ---------")
        .append(System.lineSeparator())
        .append("[+]")
        .append(e)
        .append(System.lineSeparator())

      if (temp != null) {
        if (!temp.delete()) {
          val message = "Could not delete: ${temp.absolutePath}"
          montoyaApi.logging().logToError(message)
            error.append(message)
              .append(System.lineSeparator())
        }
      }
      return ExecutorOutput(error = error.toString())
    }
  }
}